subprojects/model-core/src/main/java/org/gradle/api/internal/provider/MappingProvider.java

```diff
@@ -21,7 +21,14 @@
 import javax.annotation.Nullable;
 
 /**
- * A mapping provider that uses a transform that 1. does not use the value contents and 2. always produces a value.
+ * <p>A mapping provider that uses a transform that 1. does not use the value contents and 2. always produces a value.
```

Comment:
Since these are complex constraints which make the class user decide whether the MappingProvider is appropriate for their case, I would make them more pronounced by putting them into an actual list.

I would also add a clarification what it means by "always produces a value". I presume that this means "the transform returns a value that is not missing." Otherwise "produces a value" is a bit too generic.